src/lib/merkle_mask/masking_merkle_tree.ml

```diff
@@ -210,15 +209,206 @@ module Make (Inputs : Inputs_intf.S) = struct
+            let h = Hash.merge ~height lh rh in
+            (h, parent, lpath)
+        | _ ->
+            failwith "compute_merge_hashes: unexpected match of nodes"
+      in
+      let process_single height (self_hash, addr, path) =
+        let parent = Addr.parent_exn addr in
+        match path with
+        | `Left sibling_hash :: rest ->
+            let new_hash = Hash.merge ~height self_hash sibling_hash in
+            (new_hash, parent, rest)
+        | `Right sibling_hash :: rest ->
+            let new_hash = Hash.merge ~height sibling_hash self_hash in
+            (new_hash, parent, rest)
+        | _ ->
+            failwith "compute_merge_hashes: path is empty"
+      in
+      let converge height task =
+        let reversed, mlast =
+          List.fold task ~init:([], None)
+            ~f:(fun (processed, mprev) ((_, addr, _) as el) ->
+              match mprev with
+              | None ->
+                  (processed, Some el)
+              | Some ((_, prev_addr, _) as prev)
+                when Addr.(equal @@ sibling prev_addr) addr ->
+                  (process_pair height (prev, el) :: processed, None)
+              | Some prev ->
+                  (process_single height prev :: processed, Some el) )
+        in
+        List.rev_append reversed
+        @@ Option.(map ~f:(process_single height) mlast |> to_list)
+      in
+      let rec impl acc height task =
+        let acc' =
+          List.unordered_append (List.map ~f:(fun (a, b, _) -> (b, a)) task) acc
+        in
+        match task with
+        | [] | [ (_, _, []) ] ->
+            acc'
+        | _ ->
+            impl acc' (height + 1) (converge height task)
+      in
+      let hash_account =
+        Option.value_map ~default:Hash.empty_account ~f:Hash.hash_account
+      in
+      Fn.compose (impl [] 0) (List.map ~f:(Tuple3.map_fst ~f:hash_account))
+
+    let finalize_hashes_do t unhashed_accounts =
+      let with_merkle_path_batch accs =
+        let { hashes; _ }, ancestor = maps_and_ancestor t in
+        path_batch_impl
+          ~base_lookup:(Base.merkle_path_batch ancestor)
+          ~self_lookup:
+            (self_merkle_path ~current_location:t.current_location
+               ~depth:t.depth ~hashes )
+          ~fixup_path:(fixup_merkle_path ~hashes)
+          (List.map ~f:snd accs)
+        |> List.map2_exn accs ~f:(fun (a, loc) p ->
+               (a, Location.to_path_exn loc, p) )
+      in
+      let on_snd f (_, a) (_, b) = f a b in
+      List.stable_sort ~compare:(on_snd Location.compare) unhashed_accounts
+      |> List.remove_consecutive_duplicates ~which_to_keep:`First
+           ~equal:(on_snd Location.equal)
+      |> with_merkle_path_batch |> compute_merge_hashes
+      |> List.iter ~f:(Tuple2.uncurry @@ self_set_hash_impl t)
+
+    let finalize_hashes t =
+      let unhashed_accounts = t.unhashed_accounts in
+      if not @@ List.is_empty unhashed_accounts then (
+        t.unhashed_accounts <- [] ;
```

Comment:
Is it important to first nullify the existing `unhashed_accounts` before calling `finalize_hashes_do`?... Would the other way around work?

It's confusing because `finalize_hashes_do` also takes `t` as an argument, so you pass `unhashed_accounts` into it twice technically. `finalize_hashes_do` could just read/nullify these internally.